File: desktop-app/src-tauri/src/quota/agy_cli.rs
use std::process::Stdio;
use std::time::Duration;

use serde_json::Value;
use tokio::process::Command;

use crate::{FullStatus, QuotaData};

use super::credentials::find_agy_binary;
use super::process::hide_tokio_command;
use super::{is_usable_status, ProviderError, ProviderErrorKind};

const PROVIDER: &str = "agy CLI";
const COMMAND_TIMEOUT: Duration = Duration::from_secs(12);

#[derive(Default)]
struct WindowValue {
    percent: u32,
    reset: String,
    present: bool,
}

#[derive(Default)]
struct ProviderWindows {
    five_hour: WindowValue,
    weekly: WindowValue,
}

pub fn parse_quota_envelope(raw: &str) -> Result<FullStatus, ProviderError> {
    let root: Value = serde_json::from_str(raw)
        .map_err(|_| ProviderError::new(PROVIDER, ProviderErrorKind::InvalidData, "quota output was not valid JSON"))?;
    let data = root
        .pointer("/command/data")
        .or_else(|| root.get("data"))
        .ok_or_else(|| ProviderError::new(PROVIDER, ProviderErrorKind::InvalidData, "quota JSON did not contain command.data"))?;
    let groups = data
        .get("groups")
        .and_then(Value::as_array)
        .ok_or_else(|| ProviderError::new(PROVIDER, ProviderErrorKind::InvalidData, "quota JSON did not contain command.data.groups"))?;

    let mut gemini = ProviderWindows::default();
    let mut shared = ProviderWindows::default();
    let mut saw_recognized_group = false;

    for group in groups {
        let name = group
            .get("name")
            .or_else(|| group.get("displayName"))
            .and_then(Value::as_str)
            .unwrap_or("")
            .to_lowercase();
        let target = if name.contains("gemini") {
            saw_recognized_group = true;
            Some(&mut gemini)
        } else if name.contains("claude") || name.contains("gpt") || name.contains("openai") {
            saw_recognized_group = true;
            Some(&mut shared)
        } else {
            None
        };
        let Some(target) = target else { continue };
        let Some(buckets) = group.get("buckets").and_then(Value::as_array) else { continue };
        for bucket in buckets {
            let window = window_name(bucket);
            let Some(percent) = remaining_percent(bucket) else { continue };
            let reset = reset_time(bucket).unwrap_or_default();
            if is_five_hour_window(&window) {
                assign_window(&mut target.five_hour, percent, reset);
            } else if is_weekly_window(&window) {
                assign_window(&mut target.weekly, percent, reset);
            }
        }
    }

    if !saw_recognized_group {
        return Err(ProviderError::new(PROVIDER, ProviderErrorKind::InvalidData, "quota JSON contained no recognized provider groups"));
    }

    let plan_tier = data
        .get("planTier")
        .or_else(|| data.get("plan_tier"))
        .or_else(|| data.get("tier"))
        .and_then(Value::as_str)
        .map(str::to_string);

    let status = FullStatus {
        credits: None,
        quotas: vec![build_card("Gemini", gemini), build_card("Claude & OpenAI", shared)],
        plan_tier,
        recently_used_model: None,
    };
    Ok(status)
}

fn build_card(model: &str, windows: ProviderWindows) -> QuotaData {
    let (percent, refresh_time) = if windows.five_hour.present {
        (windows.five_hour.percent, windows.five_hour.reset.clone())
    } else if windows.weekly.present {
        (windows.weekly.percent, windows.weekly.reset.clone())
    } else {
        (0, String::new())
    };
    QuotaData {
        model: model.to_string(),
        percent,
        refresh_time,
        five_hour_percent: windows.five_hour.percent,
        five_hour_reset: windows.five_hour.reset,
        five_hour_disabled: !windows.five_hour.present,
        weekly_percent: windows.weekly.percent,
        weekly_reset: windows.weekly.reset,
        weekly_disabled: !windows.weekly.present,
    }
}

fn assign_window(slot: &mut WindowValue, percent: u32, reset: String) {
    if !slot.present || percent < slot.percent {
        slot.percent = percent;
        slot.reset = reset;
        slot.present = true;
    }
}

fn window_name(bucket: &Value) -> String {
    bucket
        .get("window")
        .or_else(|| bucket.get("name"))
        .or_else(|| bucket.get("label"))
        .or_else(|| bucket.get("duration"))
        .and_then(Value::as_str)
        .unwrap_or("")
        .trim()
        .to_lowercase()
}

fn is_five_hour_window(window: &str) -> bool {
    let compact: String = window.chars().filter(|c| !c.is_whitespace() && *c != '-' && *c != '_').collect();
    compact == "5h" || compact.contains("5hour") || compact.contains("fivehour")
}

fn is_weekly_window(window: &str) -> bool {
    window.contains("week") || window == "7d"
}

fn remaining_percent(bucket: &Value) -> Option<u32> {
    for key in ["remainingFraction", "remaining_fraction"] {
        if let Some(value) = bucket.get(key).and_then(number_value) {
            return Some((value.clamp(0.0, 1.0) * 100.0).round() as u32);
        }
    }
    for key in ["percentRemaining", "percent_remaining", "remainingPercent", "remaining_percent"] {
        if let Some(value) = bucket.get(key).and_then(number_value) {
            return Some(value.clamp(0.0, 100.0).round() as u32);
        }
    }
    None
}

fn number_value(value: &Value) -> Option<f64> {
    value.as_f64().or_else(|| value.as_str()?.parse::<f64>().ok())
}

fn reset_time(bucket: &Value) -> Option<String> {
    bucket
        .get("resetTime")
        .or_else(|| bucket.get("reset_time"))
        .and_then(Value::as_str)
        .filter(|s| !s.is_empty())
        .map(str::to_string)
}

pub async fn fetch() -> Result<FullStatus, ProviderError> {
    let binary = find_agy_binary()
        .ok_or_else(|| ProviderError::new(PROVIDER, ProviderErrorKind::Unavailable, "agy executable was not found"))?;

    let mut last_error = None;
    for command in ["/usage", "/quota"] {
        match run_structured_command(&binary, command).await {
            Ok(stdout) => match parse_quota_envelope(&stdout) {
                Ok(status) if is_usable_status(&status) => return Ok(status),
                Ok(_) => {
                    last_error = Some(ProviderError::new(PROVIDER, ProviderErrorKind::InvalidData, "agy returned an empty quota snapshot"));
                }
                Err(error) => last_error = Some(error),
            },
            Err(error) if error.kind == ProviderErrorKind::Auth || error.kind == ProviderErrorKind::Transient => {
                return Err(error);
            }
            Err(error) => last_error = Some(error),
        }
    }

    Err(last_error.unwrap_or_else(|| ProviderError::new(PROVIDER, ProviderErrorKind::Unsupported, "agy did not expose a structured quota command")))
}

async fn run_structured_command(binary: &std::path::Path, slash_command: &str) -> Result<String, ProviderError> {
    let mut command = Command::new(binary);
    hide_tokio_command(&mut command);
    command
        .args(["-p", slash_command, "--output-format", "json"])
        .stdin(Stdio::null())
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .kill_on_drop(true);

    let output = tokio::time::timeout(COMMAND_TIMEOUT, command.output())
        .await
        .map_err(|_| ProviderError::new(PROVIDER, ProviderErrorKind::Transient, "agy quota command timed out"))?
        .map_err(|_| ProviderError::new(PROVIDER, ProviderErrorKind::Unavailable, "failed to start agy"))?;

    let stdout = String::from_utf8_lossy(&output.stdout).to_string();
    let stderr = String::from_utf8_lossy(&output.stderr).to_string();
    if output.status.success() {
        if stdout.trim().is_empty() {
            return Err(ProviderError::new(PROVIDER, ProviderErrorKind::InvalidData, "agy returned empty structured output"));
        }
        return Ok(stdout);
    }

    let combined = format!("{} {}", stderr, stdout).to_lowercase();
    let kind = if combined.contains("login")
        || combined.contains("sign in")
        || combined.contains("sign-in")
        || combined.contains("auth")
        || combined.contains("credential")
    {
        ProviderErrorKind::Auth
    } else if combined.contains("unknown command")
        || combined.contains("unexpected arguments")
        || combined.contains("flags provided but not defined")
        || combined.contains("unknown flag")
        || combined.contains("not defined")
    {
        ProviderErrorKind::Unsupported
    } else if combined.contains("timeout")
        || combined.contains("network")
        || combined.contains("connection")
        || combined.contains("temporar")
    {
        ProviderErrorKind::Transient
    } else {
        ProviderErrorKind::Unsupported
    };
    Err(ProviderError::new(PROVIDER, kind, format!("agy could not run structured {slash_command}")))
}

File: src/telemetry/providers/agyCli.ts
import { spawn } from 'child_process';
import * as fs from 'fs';
import * as os from 'os';
import * as path from 'path';

import { FullStatus, QuotaData } from '../types';
import { ProviderError, ProviderErrorKind, isUsableStatus } from './types';

const PROVIDER = 'agy CLI';
const COMMAND_TIMEOUT_MS = 12_000;
const OUTPUT_LIMIT_BYTES = 1024 * 1024;
const MISSING_BINARY_TTL_MS = 5 * 60_000;

let missingBinaryUntil = 0;
let unsupportedContractKey: string | null = null;

interface WindowValue {
  percent: number;
  reset: string;
  present: boolean;
}

interface ProviderWindows {
  fiveHour: WindowValue;
  weekly: WindowValue;
}

function emptyWindow(): WindowValue {
  return { percent: 0, reset: '', present: false };
}

function emptyProviderWindows(): ProviderWindows {
  return { fiveHour: emptyWindow(), weekly: emptyWindow() };
}

function isFile(candidate: string | undefined | null): candidate is string {
  if (!candidate) return false;
  try {
    return fs.statSync(candidate).isFile();
  } catch {
    return false;
  }
}

export function findAgyBinary(): string | null {
  if (Date.now() < missingBinaryUntil) {
    return null;
  }

  const override = process.env.AGY_BIN;
  if (isFile(override)) {
    return override;
  }

  const executable = process.platform === 'win32' ? 'agy.exe' : 'agy';
  const pathValue = process.env.PATH;
  if (pathValue) {
    for (const directory of pathValue.split(path.delimiter)) {
      if (!directory) continue;
      const candidate = path.join(directory, executable);
      if (isFile(candidate)) {
        return candidate;
      }
    }
  }

  if (process.platform === 'win32') {
    const localAppData = process.env.LOCALAPPDATA;
    if (localAppData) {
      const candidate = path.join(localAppData, 'agy', 'bin', 'agy.exe');
      if (isFile(candidate)) {
        return candidate;
      }
    }
  } else {
    const home = os.homedir();
    if (home) {
      const candidate = path.join(home, '.local', 'bin', 'agy');
      if (isFile(candidate)) {
        return candidate;
      }
    }
    if (isFile('/usr/local/bin/agy')) {
      return '/usr/local/bin/agy';
    }
  }

  missingBinaryUntil = Date.now() + MISSING_BINARY_TTL_MS;
  return null;
}

function binaryContractKey(binary: string): string {
  try {
    return `${binary}:${fs.statSync(binary).mtimeMs}`;
  } catch {
    return binary;
  }
}

function numberValue(value: unknown): number | null {
  if (typeof value === 'number' && Number.isFinite(value)) return value;
  if (typeof value === 'string' && value.trim()) {
    const parsed = Number(value);
    return Number.isFinite(parsed) ? parsed : null;
  }
  return null;
}

function remainingPercent(bucket: any): number | null {
  for (const key of ['remainingFraction', 'remaining_fraction']) {
    const value = numberValue(bucket?.[key]);
    if (value !== null) {
      return Math.round(Math.max(0, Math.min(1, value)) * 100);
    }
  }

  for (const key of [
    'percentRemaining',
    'percent_remaining',
    'remainingPercent',
    'remaining_percent',
  ]) {
    const value = numberValue(bucket?.[key]);
    if (value !== null) {
      return Math.round(Math.max(0, Math.min(100, value)));
    }
  }

  return null;
}

function resetTime(bucket: any): string {
  const value = bucket?.resetTime ?? bucket?.reset_time;
  return typeof value === 'string' ? value : '';
}

function windowName(bucket: any): string {
  const value = bucket?.window ?? bucket?.name ?? bucket?.label ?? bucket?.duration;
  return typeof value === 'string' ? value.trim().toLowerCase() : '';
}

function isFiveHourWindow(window: string): boolean {
  const compact = window.replace(/[\s_-]/g, '');
  return compact === '5h' || compact.includes('5hour') || compact.includes('fivehour');
}

function isWeeklyWindow(window: string): boolean {
  const compact = window.replace(/[\s_-]/g, '');
  return compact.includes('week') || compact === '7d';
}

function assignWindow(slot: WindowValue, percent: number, reset: string): void {
  if (!slot.present || percent < slot.percent) {
    slot.percent = percent;
    slot.reset = reset;
    slot.present = true;
  }
}

function buildCard(model: string, windows: ProviderWindows): QuotaData {
  const legacy = windows.fiveHour.present
    ? windows.fiveHour
    : windows.weekly.present
      ? windows.weekly
      : emptyWindow();

  return {
    model,
    percent: legacy.percent,
    refreshTime: legacy.reset,
    fiveHourPercent: windows.fiveHour.percent,
    fiveHourReset: windows.fiveHour.reset,
    fiveHourDisabled: !windows.fiveHour.present,
    weeklyPercent: windows.weekly.percent,
    weeklyReset: windows.weekly.reset,
    weeklyDisabled: !windows.weekly.present,
  };
}

export function parseAgyQuotaEnvelope(raw: string): FullStatus {
  let root: any;
  try {
    root = JSON.parse(raw);
  } catch {
    throw new ProviderError(
      PROVIDER,
      'invalid-data',
      'quota output was not valid JSON',
    );
  }

  const data = root?.command?.data ?? root?.data;
  if (!data || !Array.isArray(data.groups)) {
    throw new ProviderError(
      PROVIDER,
      'invalid-data',
      'quota JSON did not contain structured groups',
    );
  }

  const gemini = emptyProviderWindows();
  const shared = emptyProviderWindows();
  let sawRecognizedGroup = false;

  for (const group of data.groups) {
    const rawName = group?.name ?? group?.displayName ?? '';
    const name = typeof rawName === 'string' ? rawName.toLowerCase() : '';

    let target: ProviderWindows | null = null;
    if (name.includes('gemini')) {
      target = gemini;
      sawRecognizedGroup = true;
    } else if (
      name.includes('claude') ||
      name.includes('gpt') ||
      name.includes('openai')
    ) {
      target = shared;
      sawRecognizedGroup = true;
    }

    if (!target || !Array.isArray(group?.buckets)) continue;

    for (const bucket of group.buckets) {
      const percent = remainingPercent(bucket);
      if (percent === null) continue;
      const reset = resetTime(bucket);
      const window = windowName(bucket);
      if (isFiveHourWindow(window)) {
        assignWindow(target.fiveHour, percent, reset);
      } else if (isWeeklyWindow(window)) {
        assignWindow(target.weekly, percent, reset);
      }
    }
  }

  if (!sawRecognizedGroup) {
    throw new ProviderError(
      PROVIDER,
      'invalid-data',
      'quota JSON contained no recognized provider groups',
    );
  }

  const tierCandidate = data.planTier ?? data.plan_tier ?? data.tier;
  const planTier = typeof tierCandidate === 'string' && tierCandidate
    ? tierCandidate
    : null;

  return {
    credits: null,
    quotas: [
      buildCard('Gemini', gemini),
      buildCard('Claude & OpenAI', shared),
    ],
    recentlyUsedModel: null,
    planTier,
  };
}

function classifyFailure(text: string): ProviderErrorKind {
  const lower = text.toLowerCase();
  if (
    lower.includes('login') ||
    lower.includes('sign in') ||
    lower.includes('sign-in') ||
    lower.includes('auth') ||
    lower.includes('credential')
  ) {
    return 'auth';
  }
  if (
    lower.includes('unknown command') ||
    lower.includes('unexpected arguments') ||
    lower.includes('flags provided but not defined') ||
    lower.includes('unknown flag') ||
    lower.includes('not defined')
  ) {
    return 'unsupported';
  }
  if (
    lower.includes('timeout') ||
    lower.includes('network') ||
    lower.includes('connection') ||
    lower.includes('temporar')
  ) {
    return 'transient';
  }
  return 'unsupported';
}

function runStructuredCommand(binary: string, slashCommand: string): Promise<string> {
  return new Promise((resolve, reject) => {
    const child = spawn(
      binary,
      ['-p', slashCommand, '--output-format', 'json'],
      {
        windowsHide: true,
        stdio: ['ignore', 'pipe', 'pipe'],
      },
    );

    let stdout = '';
    let stderr = '';
    let stdoutBytes = 0;
    let stderrBytes = 0;
    let settled = false;

    const finishReject = (error: ProviderError) => {
      if (settled) return;
      settled = true;
      clearTimeout(timer);
      reject(error);
    };

    const timer = setTimeout(() => {
      child.kill();
      finishReject(
        new ProviderError(PROVIDER, 'transient', 'agy quota command timed out'),
      );
    }, COMMAND_TIMEOUT_MS);

    child.stdout?.on('data', (chunk: Buffer) => {
      if (settled) return;
      stdoutBytes += chunk.length;
      if (stdoutBytes > OUTPUT_LIMIT_BYTES) {
        child.kill();
        finishReject(
          new ProviderError(PROVIDER, 'invalid-data', 'agy stdout exceeded output limit'),
        );
        return;
      }
      stdout += chunk.toString('utf8');
    });

    child.stderr?.on('data', (chunk: Buffer) => {
      if (settled) return;
      stderrBytes += chunk.length;
      if (stderrBytes > OUTPUT_LIMIT_BYTES) {
        child.kill();
        finishReject(
          new ProviderError(PROVIDER, 'invalid-data', 'agy stderr exceeded output limit'),
        );
        return;
      }
      stderr += chunk.toString('utf8');
    });

    child.on('error', () => {
      finishReject(
        new ProviderError(PROVIDER, 'unavailable', 'failed to start agy'),
      );
    });

    child.on('close', (code) => {
      if (settled) return;
      settled = true;
      clearTimeout(timer);

      if (code === 0) {
        if (!stdout.trim()) {
          reject(
            new ProviderError(
              PROVIDER,
              'invalid-data',
              'agy returned empty structured output',
            ),
          );
          return;
        }
        resolve(stdout);
        return;
      }

      const kind = classifyFailure(`${stderr} ${stdout}`);
      reject(
        new ProviderError(
          PROVIDER,
          kind,
          `agy could not run structured ${slashCommand}`,
        ),
      );
    });
  });
}

export async function fetchAgyCli(force: boolean): Promise<FullStatus> {
  void force;

  const binary = findAgyBinary();
  if (!binary) {
    throw new ProviderError(PROVIDER, 'unavailable', 'agy executable was not found');
  }

  const contractKey = binaryContractKey(binary);
  if (unsupportedContractKey === contractKey) {
    throw new ProviderError(
      PROVIDER,
      'unsupported',
      'installed agy does not expose structured quota output',
    );
  }

  let lastError: ProviderError | null = null;
  for (const slashCommand of ['/usage', '/quota']) {
    try {
      const stdout = await runStructuredCommand(binary, slashCommand);
      const status = parseAgyQuotaEnvelope(stdout);
      if (isUsableStatus(status)) {
        return status;
      }
      lastError = new ProviderError(
        PROVIDER,
        'invalid-data',
        'agy returned an empty quota snapshot',
      );
    } catch (error) {
      const providerError = error instanceof ProviderError
        ? error
        : new ProviderError(PROVIDER, 'unsupported', 'agy quota command failed');
      if (providerError.kind === 'auth' || providerError.kind === 'transient') {
        throw providerError;
      }
      lastError = providerError;
    }
  }

  if (
    lastError?.kind === 'unsupported' ||
    lastError?.kind === 'invalid-data'
  ) {
    unsupportedContractKey = contractKey;
  }

  throw lastError ?? new ProviderError(
    PROVIDER,
    'unsupported',
    'agy did not expose a structured quota command',
  );
}
